### PR TITLE
checkout: Support a "pure addition" mode

### PR DESCRIPTION
**Pull request by cgwalters**
I plan to use this for `rpm-ostree livefs`.
https://github.com/projectatomic/rpm-ostree/issues/639

### CODE REVIEW

**Comment by cgwalters** [comment]
WIP due to lack of tests, cmdline wrapper.

**Comment by cgwalters** [comment]
Now with tests ✅ - removing WIP.

File: src/libostree/ostree-repo-checkout.c [hunk 1]
```diff
@@ -362,6 +375,12 @@ checkout_file_hardlink (OstreeRepo                          *self,
     {
       ret_was_supported = FALSE;
     }
+  else if (errno == EEXIST && options->overwrite_mode == OSTREE_REPO_CHECKOUT_OVERWRITE_ADD_FILES)
+    {
+      /* In this mode, we keep existing content.  Still need to mark the hardlink as supported.
+       */
+      ret_was_supported = TRUE;
```
Comment (jlebon) [inline src/libostree/ostree-repo-checkout.c]: One concern with this is that the kernel checks for `EEXISTS` before checking whether hardlinking is supported. So we can't really assert this, right? Looking at `checkout_one_file_at()`, it relies on this info to update the devino cache, which will be wrong after this. Maybe add e.g. `out_exists` also so callers can distinguish this?

Comment (cgwalters) [inline src/libostree/ostree-repo-checkout.c]: Oooh.  That's a good catch.  That would have been spectacularly painful to debug later.

**Comment by cgwalters** [comment]
Ok, rebased 🏄, added a prep patch and a fixup.

File: src/libostree/ostree-repo-checkout.c [hunk 2]
```diff
@@ -579,14 +581,14 @@ checkout_one_file_at (OstreeRepo                        *repo,
...
-                                   FALSE, &did_hardlink,
+                                   FALSE, &hardlink_res,
                                    cancellable, error))
         {
           g_prefix_error (error, "Using new cached uncompressed hardlink of %s to %s: ", checksum, destination_name);
           goto out;
         }
 
-      need_copy = !did_hardlink;
+      need_copy = hardlink_res == HARDLINK_RESULT_NOT_SUPPORTED;
```
Comment (jlebon) [inline src/libostree/ostree-repo-checkout.c]: Parentheses on these, please!

Comment (cgwalters) [inline src/libostree/ostree-repo-checkout.c]: Done ⬇️

**Comment by jlebon** [comment]
Nice! Looks like some tests might need to be adjusted to grep for the new error message.

**Comment by cgwalters** [comment]
Eh, my fixups went wrong, I just squashed to simplify.  One of the bugs was we need to initialize `hardlink_res`, since it's not set if we're not going to try to hardlink.

**Comment by jlebon** [comment]
@rh-atomic-bot r+ 0188b3c

**Pull request closed without merging**